File: OpenMP_part1/mult.f90
program mult
implicit none
  !! Program matrix multiplication; C = AB ,
  !! where A is mxn, and B is nxp and C is A mxp matrix
  integer::i,j,k
  real:: tmp
  integer::m,n,p,size
  real::aval,bval,tol

  real, allocatable, dimension (:,:):: A,B,C
  real:: cval,err,errsq
  real(8)::wstart
  real(8)::omp_get_wtime
  integer:: omp_get_num_threads, omp_get_thread_num

  aval=3.0
  bval=5.0
  tol =0.001
  size = 2000
  m=size
  n=size
  p=size

  allocate (A(n,p))
  allocate (B(p,m))
  allocate (C(n,m))

  ! Init A,B and C matrices
  A(:,:)=aval
  B(:,:)=bval
  C(:,:)=0.0

  wstart = omp_get_wtime();

  ! Calculate the C matrix
  !$omp parallel do private(tmp)
  do i=1,n
    do j=1,m
      tmp=0.0
      do k=1,p
        tmp = tmp + A(i,k) * B(k,j)
      end do
      C(i,j)=tmp
    end do
  end do
  !$omp end parallel do

  write(6,*) 'time ',omp_get_wtime()-wstart

  ! Error check
  cval = p * aval* bval
  errsq = 0.0

  do i=1, n
    do j=1, m
      err = C(i,j) - cval;
      errsq = errsq + err*err;
    end do
  end do

  if(errsq > tol) then
    write(6,*)'Error in the multiplication:' ,errsq
  else
    write(6,*)'It works'
  end if

  deallocate(A)
  deallocate(B)
  deallocate(C)

end program mult

File: OpenMP_part1/mult.c
#include "omp.h"
#include <stdio.h>
#include <malloc.h>
#define AVAL 3.0
#define BVAL 5.0
#define TOL 0.001

// Program: Matrix multiplication C=AB,
// where A is mxn, B is nxp and C is mxp matrix.

int main(int argc,char *argv[])
{
  
  int i,j,k;

  double tmp=0.0;

  int size = 2000; //Try different sizes
  int n = size;
  int m = size; 
  int p = size;
  
 // double A[n][p];
 // double B[p][m];
 // double C[n][m];
  //Dynamic memory allocation
  double **A = (double**) malloc(sizeof(double*)*n);
  double **B = (double**) malloc(sizeof(double*)*p);
  double **C = (double**) malloc(sizeof(double*)*n);
  for (i=0;i<n;i++)
  {
     A [i] = (double*) malloc(sizeof(double)*p);
     C [i] = (double*) malloc(sizeof(double)*m);
  }
  for (j=0;j<p;j++)
    B [j] = (double*) malloc(sizeof(double)*m);
 

  double cval,err,errsq;

  /*Init A,B and C matrices*/
  for (i=0;i<n;i++)
    for (j=0;j<p;j++)
      A[i][j] = AVAL;

  for (j=0;j<p;j++)
    for (k=0;k<m;k++)
      B[j][k] = BVAL;

  for (i=0;i<n;i++)
    for (j=0;j<m;j++)
      C[i][j] = 0.0;
 
  
  double wstart = omp_get_wtime();

  /*Calculate the C matrix  (Something to do here?)*/
  
 for (i=0;i<n;i++)
  {
    for (j=0;j<m;j++)
    {

      tmp=0.0;
      for (k=0;k<p;k++)
      {
	 tmp += A[i][k] * B[j][k];
      }
      C[i][j]=tmp;    
  
    }
  }

  printf("time %f\n",omp_get_wtime()-wstart);

  /* Error check*/
  cval = p * AVAL* BVAL;
  errsq = 0.0;

  for(i=0;i<n;i++)
    for (j=0;j<m;j++)
    {
      err = C[i][j]-cval;
      errsq += err*err;
    }

  if(errsq > TOL)
    printf("\n Error in the multiplication: %f\n",errsq);
  else
    printf("\n Its work %f\n",cval);
  return 0;
}
